### DISCUSSION TITLE
ESP32 Audio Kit v2.2 2-way audio

### DISCUSSION THREAD

**Discussion by W7SVT** [pschatzmann/arduino-audio-tools#641] [Q&A] [answered]
I'm trying to get Sink and Source working over BT Classic on the ESP32 Audio Kit v2.2 and can't seem to make much progress. Has anyone gotten this to work? Working on some Home automation and am trying to replicate Alexa Drop in basic functionality

**Comment by pschatzmann** [comment]
I am not sure what functionality are you referring to exactly ?
If you want to send audio over bluetooth serial you will need to make sure that the amount of data you generate is fitting with the transmission capacity by reducing the sample rate, channels and/or using a codec

**Reply by W7SVT** [reply]
The code will probably make it more clear and what I should have posted to begin with

#include <BluetoothA2DPSink.h>
#include <BluetoothA2DPSource.h>
#include <Audio.h>
#include <Wire.h>
#include <SPI.h>
#include <SD.h>

// Initialize the BluetoothA2DPSink and BluetoothA2DPSource objects
BluetoothA2DPSink a2dp_sink;
BluetoothA2DPSource a2dp_source;

// Initialize the Audio object
Audio audio;

void setup() {
  // Start the serial communication
  Serial.begin(9600);
  
  // Initialize the SD card
  if (!SD.begin()) {
    Serial.println("SD card initialization failed!");
    return;
  }
  
  // Initialize the Audio library
  audio.begin();
  
  // Configure the audio input and output
  audio.setPinout(I2S_BCLK, I2S_LRC, I2S_DOUT, I2S_DIN);
  audio.setVolume(5);

  // Initialize the BluetoothA2DPSink and BluetoothA2DPSource objects
  a2dp_sink.begin("ESP32-Audio-Kit"); // set the Bluetooth device name
  a2dp_source.begin("ESP32-Audio-Kit"); // set the Bluetooth device name

  Serial.println("Setup complete");
}

void loop() {
  // Check for a new audio data packet from the phone
  if (a2dp_sink.available()) {
    // Read the audio data and play it through the audio output channel
    audio.play(a2dp_sink.read());
  }
  
  // Check for a new audio data packet from the microphone
  if (audio.isRecording()) {
    // Read the audio data and transmit it over Bluetooth to the computer
    a2dp_source.write(audio.record());
  }
}

**Reply by W7SVT** [reply]
I'm not a slave to this idea, I just want to use the audio kit to send and receive audio over Bluetooth to a smartphone

**Comment by pschatzmann** [comment]
Some people seem to have issues with the stream based implementation (mainly as an audio source).  
I suggest to use the examples from [examples-basic-api](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-basic-api)

Just replace the I2SStream with an AudioKitStream...

**Reply by W7SVT** [reply]
I pulled inspiration from

https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/basic-a2dp-i2s/basic-a2dp-i2s.ino

Did I miss the point?

**Comment by pschatzmann** [comment, marked as answer]
Just had a look at your sketch and there are many problems with it
- You can can only use either a A2DP sink or a source in one sketch but never both
- You forgot to implement the callbacks

So using A2DP for a 2Way Radio will not work!. I suggest to use ESP-Now instead or a TCP/IP based protocol instead...

**Reply by W7SVT** [reply]
You are correct. Regarding the callbacks, they are not in the snippet I sent as their implementation is on another computer. That said, I did not know that I could only you use the line in or the headphone jack. I specifically chose this board with the hopes of doing both. Somehow I just assumed that if a $3 Bluetooth headset could do it that a bespoked and exquisitely designed dev board would be able to. Thank you for your patience and assistance.

**Comment by pschatzmann** [comment]
What makes you think that you need to can only you use the line in or the headphone jack ?
You can use both at the same time! The issue is with A2DP which can only implement one scenario at a time...

The only issue with this board (if you have a ES8388 version) however is, that you can't switch off the microphones.

**Reply by W7SVT** [reply]
I must have misunderstood what you meant about "You can can only use either a A2DP sink or a source in one sketch but never both". I took that to mean that I could only use one of the other at a time. How should I proceed to my desired end of enabling 2-way audio from a smart phone to a ESP32 Audio Kit v2.2 without ESPNOW because it would require a more then a paired phone (at least 3 ESP32 in my experience)? or TCP/IP
